Channel participation API without system channel (#24)
hyperledger/fabric-rfcs: text/0000-channel-participation-api-without-system-channel.md

---
layout: default
title: Channel participation api without system channel
nav_order: 3
---

- Feature Name: channel_participation_api_without_system_channel
- Start Date: 2020-03-01
- RFC PR: (leave this empty)
- Fabric Component: orderer
- Fabric Issue: (leave this empty)

# Summary
[summary]: #summary

Currently in order to create a channel one needs to issue a config transaction to the system channel. This has 
privacy, scalability and operational disadvantages, since the system channel (and the orderers running it) is aware of 
all channels and of all channel members (at creation time). In this feature we propose to expose a 
"channel participation" API that would allow a local orderer administrator to join and leave a channel, as well as 
to list all the channels that the local orderer is part of. This allows a Fabric network to be operated without the use 

> of a system channel, which improve the privacy and scalability of the ordering service and the network as a whole. 

Reviewer: ```suggestion
of a system channel, which improves the privacy and scalability of the ordering service and the network as a whole. 
```

Author: Fixed

# Motivation
[motivation]: #motivation

Creating channels by issuing a transaction to the system channel presents several disadvantages.

## Privacy problems

> A system channel requires that all ordering service nodes (OSNs) be aware of all channels. As Fabric depricated the 

Reviewer: ```suggestion
A system channel requires that all ordering service nodes (OSNs) be aware of all channels. As Fabric deprecated the 
```

Author: Fixed

> Kafka-based ordering service and moved to Raft (and later is planned to introduce BFT consensus), it is no longer 

Reviewer: ```suggestion
Kafka-based ordering service and moved to Raft, it is no longer 
```

No need to mention any future stuff here IMHO

Author: Fixed

desirable that an OSN that services a channel `A` with member organizations `X & Y` be aware of channel `B` with member 
organizations `Y & Z`. This creates a privacy problem, because the orderer organization (or organizations) now knows 
that `Y`, `X`'s partner to channel `A`, is also doing business with organization `Z`.  

There are workarounds that reduce the privacy problem, but not eliminate it. For example, using the system channel it 
is possible to create an application channel with only a single org, and then add additional orgs by updating the 
channel itself. However, these type of solutions are cumbersome and only partial.

## Scalability problems
All OSNs are members of the system channel, which creates a scalability problem in a large scale Fabric network. When 
the number of channels increases, Raft allows us do decouple the consenters sets of the different application channels, 
achieving linear horizontal scalability in the number of channels. That is, more resource can be added as the number of 

> channels increase. However, the system channel is an exception: decoupling application channel consenters sets as 

Reviewer: ```suggestion
channels increase. However, the system channel is an exception; decoupling application channel consenters sets as 
```

Author: Fixed

described above will cause its number of members to increase, and hence its performance to decrease. Joining and 
leaving channels without a system channel solves this problem.     
   
Moreover, in order to start a new orderer and have it join an existing channel (perform 'on-boarding'), a new orderer 
starts by scanning the system channel in order to discover existing channels and their membership. As the number of 
channels increase, this scan prolongs the process of adding a new orderer.  
    
## Operational problems
There are several operational problems that affect large scale deployment and hinder efficient management of cloud 
environments using automatic tooling.
     
Channel creation (using the system channel) utilizes information contained in the system channel - e.g. MSP information 

> of the organization - in order to apply it onto the newly created channel. This create the convoluted situation in 

Reviewer: ```suggestion
of the organization - in order to apply it to the newly created channel. This creates the convoluted situation in 
```

Author: Fixed

which an organization that wants to update its details in the consortium must coordinate that action with third parties 
that have write privileges to the system channel.

There is no good way to list all the channels that an orderer is a member of. 

There is no good way to dispose of the resources associated with a channel, once an orderer is no longer a member of 
the consenters set of a that given channel.

In many cloud environments, it is necessary to restart an orderer in order for it to discover immediately that it has 

> been added to a channel it was not a part of previously, because otherwise detection takes minutes. 

Reviewer: These should probably be a bulleted list for better readability

Author: Done

   
# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

> We introduce a new mode of channel participation management that does not depend on the existence of a system channel.

Reviewer: ```suggestion
We introduce a new mode of channel management that does not depend on the existence of a system channel.
```

Author: Done

We introduce the ability to start an OSN process without any channels. The OSN exposes a new **"channel participation" 
API** that allows the operator to join and leave channels on that specific OSN.   

> A **"local orderer channel participation administrator"** is a role that is associated with the operation and 

Reviewer: Do we really need to be fine-grained here or is a local orderer admin fine (similar to a local peer admin)?

Author: For the sake of generality, I think it is useful to formally define this role. 
Our current design choice is to have this role will use the same AA mechanism as the "operations API" admin (see Security section). However it is conceivable that these roles would be separated some day.

administration of an OSN. The identity associated with this role is able to invoke a new **"channel participation" 
API**, and perform the following commands:

- **Join** this OSN to a new or an existing channel,
- **Remove** this OSN from an existing channel,
- **List** the channels this OSN is a member of.   

The new mode of channel participation management is meant to be incompatible with channel creation using a system 

> channel, in the sense that if the system channel exists, the `Join` operation will not be supported. 

Reviewer: Does this have to be the case?  Seems like it should be possible to support both ... but I'll read about "migration" first

Author: "Mixed mode" is a problem... things can stop working in ugly ways, e.g. on-boarding. There are solutions of course, but they are complicated and costly. Therefore, for simplicity, I think we should avoid that and encourage users to use either one or the other. 

In addition, moving from system-channel mode to participation-api style is easy (delete system-channel on all OSNs). The reverse, going back to system-channel style once there are already app-channels  is complicated. Here too we choose simplicity and do not allow the reverse path. The only exception is bootstrap: starting from empty OSNs and creating the system channel.

However, there is a transition path from a network operated with a system channel to a network that is operated without

> one. In addition, it should be possible to start OSN processes without channels, create a system channel with the

Reviewer: ```
... it should be possible to start OSN processes without channels, create a system channel with the
```
This is an important point ... let's make sure it does not get lost

Author: I split the paragraph to make this more prominent.

channel participation API, and continue to operate the network in the usual way - creating channels through the system 
channel. "Mixed mode" management, that is, using the system channel to create channels on some OSNs and the 
participation API on other OSNs is not supported and highly discouraged.  

## Channel participation commands 
In this section channel participation commands are described in a high level. The commands are presented in REST style
but some details (i.e. exact formats, responses) are omitted for clarity.
   
### Joining a channel
A channel is joined by providing the OSN with a config block:

  ```Join channel-name channel-config-block```
  
  or
  

>   `POST /channels/channel-name` with body containing `channel-config-block`

Reviewer 2: It's a detail but can we add a small section on API versioning please? Resource versioning? Accept headers?

Or are we simply using REST as nomenclature instead of the target implementation pattern?

Author: I was planing (hoping...) to go into the fine-grain details when I write the Jira item dedicated to the REST API. Here it is meant to be just a sketch, in order not to confuse the reader with too many details.

  
The body contains a config block that contains all the details about the channel. 

> If the OSN is joining a new channel, the config block is the genesis block of the channel, generated by the 
> `configtxgen` tool, as described below. It looks like the first block of an 

Reviewer: ```suggestion
If the OSN is joining a new channel, the config block is the genesis block of the channel. It looks like the first block of an 
```
we don't need to say how the genesis block is created here as it's not relevant

Author: Done

application channel that is generated by the OSN, when a channel creation transaction is submitted to the system channel.

If the OSN is joining an existing channel, the config block is simply the last config block of the channel, retrieved 
from one of the other OSNs.

If the channel already exists, the call will fail.

If the system channel exists, the call will fail. 
          
The OSN receiving this call will inspect the block and conduct some sanity checks in order to make sure that the block 
is well-formed.

The OSN will try to find itself in the consenters set, and determine if it is a follower or a member of the cluster

> executing consensus. 

Reviewer 2: The roles of "follower" and "member" should be more formally defined. It's referenced here with assumptions about the reader's definition of those words which may not match the intent.

I realize the term was emphasized earlier with an implied definition but I believe an explicit definition of the concept would make this more straightforward.

Author: Agreed. I will improve the presentation.

If it is not in the set, it will continue as a **follower**, and start to pull blocks from other OSNs. It will continue
to do so until it catches up with the config block received in the invocation. From then on, it will continue to pull blocks, 
but will inspect each incoming config block and check whether it is included in the consenters set. If it is, it will 
start the runtime components that execute consensus, thus joining the cluster. 

> If the OSN is in the consenters set, it will continue as **member**, first pulling blocks from other OSMs util it catches

Reviewer: ```suggestion
If the OSN is in the consenters set, it will continue as **member**, first pulling blocks from other OSNs util it catches
```

Author: Fixed

up with the given config block, and then starting the runtime components that execute consensus, thus joining the cluster.

Note that joining an OSN as a member means that the channel was updated to include that OSN in the consenters set before it
was joined by the participation API. Therefore, during the catch-up process and until the new OSN starts executing 
consensus, the cluster is a state of reduced fault tolerance. For long chains this might be a problem, since catch-up
can take a long time. The solution is to join the new OSN as a follower. After it catches up with the cluster the 

> channel admin can update the channel config and include it in the consenters set.

Reviewer: ```
The solution is to join the new OSN as a follower. After it catches up with the cluster the 
channel admin can update the channel config and include it in the consenters set.
```
This is actually a new feature as well, correct?

Author: Yes, this is a new feature.

The new OSN can pull blocks from other OSNs only if it has authorization to do so. A minimal requirement is that its 

> owning org will be defined in the channel configuration, and that the signature of the new OSN on the `Deliver` 

Reviewer: ```suggestion
owning organization will be defined in the channel configuration, and that the signature of the new OSN on the `Deliver` 
```

Author: Fixed

> request satisfy the `/Channel/Readers` policy.

Reviewer: ```suggestion
request satisfies the `/Channel/Readers` policy.
```

Author: Fixed

#### Creating a channel genesis block
The way to create a channel genesis block is by using the `configtxgen`  tool, which already supports the 
generation of an **application** channel genesis block.

```configtxgen -outputBlock channel_genesis_block.pb -profile SampleRaftChannelCreation -channelID my-app-channel```

Generating the channel genesis block follows the same rationale as generating the system channel genesis block, when 
bootstrapping a new Fabric network. 

The command refers to a profile `SampleRaftChannelCreation` in a `configtx.yaml` file, as usual. The YAML file 
will point to the file-system path of the MSP configuration of each org, for application as well as ordering orgs. 
In addition, it will point to the path of the certificates of the consenters. Generating a genesis block for an 
application channel means that:
- the `configtx.yaml` **does not** include a `Consortiums` config group

> - the `configtx.yaml` **must include** a `Channel/Application` config group

Reviewer 2: `configtx.yaml` should not be explicitly referenced here. It's a (poor) means to an end.

Author: Done, I demoted that to an example.

 

The minimally required elements should include enough details for the OSN to achieve consensus and allow further 

> configuration of all the config groups.           

Reviewer: You should not mention `configtxgen` here ... it's just a tool.
I think the important thing to detail here is what needs to be in a channel config and the fact that there is no longer a consortium defined.  This means that you can no longer simply reference MSP content but must explicitly include it in the channel create / config transaction / genesis block.

At least I think that's the case

Author: I demoted the reference to configtxgen to a single example, rather then making it a prescription for generating the genesis block. I also described the changes w.r.t. a consortium & consortiums.

- Start one or more OSNs without any channels. 
- These OSNs should be configured to accept channel participation commands from their respective local orderer channel 
  participation admins. This is achieved by setting the respective fields in the orderer.yaml file of each OSN.
- These OSNs may belong to different organizations.
- Generate (using configtxgen, as usual), the genesis block of a system channel. (The system channel contains the 
  `Consortiums` config group, whereas application channels do not).
- Create the system channel by invoking the "Join" command as described above, but supply the genesis block of the 
  system channel, rather than that of an application channel.
- The channel participation command "Join" will no longer accept commands, and channel creation can only 
  be done using a config transaction on the system channel.
- The channel participation commands "List" and "Delete" (of system channel only) will continue to function.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## The API layer
We intend to extend the "operations API" (package `core/operations`) by adding a handler for channel participation 

> commands, for path `/channels` as described above. These commands will only be enabled if specified explicitly in the `orderer.yaml` 

Reviewer 2: `/channels` should probably be scoped to your API instead of hung off the root.

Author: You mean something like this?

`channel-participation/channels`

Author: Done

configuration file. These commands will only be available in the orderer and not in the peer.

> 

Reviewer 2: There is no "operations API". There's an embedded server that this can hook into to provide the orderer API. We must keep them separate.

The implementation also does *not* belong in core/operations.

Author: Most of the implementation will be under orderer/... 
Some tiny bits that hook into the HTTP server might reside elsewhere (not sure where though).

   - add the new OSN to the channel (join with last-config-block)
   - as a follower (not in consensus set), adding it after catch-up
   - as a member (in consensus set)
4. Remove an OSN from a channel:
   - start with a set of OSNs, with a channel, with blocks
   - add the new OSN to the channel 
        - as a follower 
        - as a member
5. Bootstrap to system channel:
   - starting a set of empty OSNs 
   - joining all to a system-channel
   - create a channel
   - make sure Join and Delete (app-channel) are disabled
5. Transition from system-channel to local participation:
   - start from a network with a system channel and app channel
   - remove system-channel
   - add channel with participation API

# Dependencies
[dependencies]: #dependencies

## Existing items
These Jira items are precursors of this RFC:

- https://jira.hyperledger.org/browse/FAB-15709
- https://jira.hyperledger.org/browse/FAB-15710
- https://jira.hyperledger.org/browse/FAB-15711
- https://jira.hyperledger.org/browse/FAB-15712
- https://jira.hyperledger.org/browse/FAB-15713

These will be combined in the development plan once the RFC is approved.

## Authorization and Authentication

> There is an ongoing effort to improve the AA mechanisms of the "operations" API. 

Reviewer 2: Where is that effort taking place? I'm very interested as it has larger impacts and is explicitly an area that I have opinions on.

Author: Gari is doing some exploratory work on that, I believe. 
@mastersingh24

administration of an OSN. The identity associated with this role is able to invoke a new **"channel participation" 
API**, and perform the following commands:

- **Join** this OSN to a new or an existing channel,
- **Remove** this OSN from an existing channel,
- **List** the channels this OSN is a member of.   

The new mode of channel participation management is meant to be incompatible with channel creation using a system 
channel, in the sense that if the system channel exists, the `Join` operation will not be supported. 
However, there is a transition path from a network operated with a system channel to a network that is operated without
one: removing the system channel from all OSNs.  

It should be possible to start OSN processes without channels, create a system channel with the
channel participation API, and continue to operate the network in the usual way - creating channels through the system 
channel. 

"Mixed mode" management, that is, using the system channel to create channels on some OSNs and the 
participation API to create channels on other OSNs is not supported and highly discouraged.  

## Channel participation commands 

> In this section channel participation commands are described in a high level. The commands are presented in REST style

Reviewer 3: Any specific reason to use REST rather than the existing gRPC style?

Author: We want to make it easier for cloud operators to manage the OSN channel participation. This way they won't need additional tools that implement gRPC, and can use a browser, for example.

Reviewer: I'm a bit torn on this one ... I tend to think we should probably stick with gRPC here.  But I will not hold things up ... the APIs generally look ok and it's not hard to translate these to protobuf defs.

Configuration changes can cause an OSN to switch between the follower and member roles. 
 
* A follower evaluates incoming config blocks and checks whether it was added to the consenters set. If it was added, 
  it starts the runtime components that execute consensus, thus becoming a member the cluster.
* A member evaluates incoming config blocks and checks whether it was removed from the consenters set. If it was 
  removed, it stops the runtime components that execute consensus, but continues to operate as a cluster follower.
     
An OSN operating as a cluster follower is a new feature introduced by this work.

#### Two ways to join     
When an OSN joins a channel as a **follower**, it start to pull blocks from other OSNs. It will continue
to do so until it **catches up** with the config block received in the invocation. From then on, it will continue to pull 
blocks, but will inspect each incoming config block and check whether it is included in the consenters set. 
When channel admins add it to the consenters set, it will detect that and start the runtime components that execute 
consensus, thus becoming a member the cluster. 

> When an OSN joins a channel as a **member**, it first pulls blocks from other OSNs util it **catches

Reviewer 4: Very minor typo -- "util" instead of "until"

Author: Fixed.

via a Go interface, and executed the operations described above.  
  
## The API layer
The peer and the orderer host an HTTP server that offers a RESTful “operations” API (see: 
[The Operations Service](https://hyperledger-fabric.readthedocs.io/en/release-2.0/operations_service.html)).
We intend to hook into the HTTP server that is hosted by the orderer, and expose a new capability 
`/channel-participation`, which has a single resource  `/channels` as described above. 
The channel participation commands will only be enabled if specified explicitly in the `orderer.yaml` 
configuration file. These commands will only be available in the orderer and not in the peer.

The _local orderer participation admin_  role will have the same authentication and authorization
as the _operations admin_ (i.e. the same mutual TLS certificate). We choose this option because it is the simplest 
to implement at this time. This is expected to change in the future (see below).

## Participation management
We intend to implement the channel participation management functionality in a new package 

> (`orderer/common/channelparticipation`). In this package a new class will keep a map of channel names and their status,

Reviewer 4: This follows the existing patterns of the orderer codebase, but we have been moving towards placing new code not to be consumed by external parties as a library (though this is essentially all of Fabric) into `./internal`.  I'd welcome opinions from other maintainers, but either way, I don't see the location of this package as being significant enough to hold up the RFC.

and will be in charge of coordinating the operations of following or joining a channel as a member. 
Every channel that needs to follow or catch-up on the cluster will have a dedicated goroutine. When the channel reaches

> a state where it can join the cluster, the respective methods in `orderer/common/multichannel/registrar.go` will be 

Reviewer 4: Being the primary author of much of the multichannel code, I would not get too hung up on preserving our existing channel startup and management structure.  If this code needs to be gutted and replaced, all the better.  Regardless, this seems like an implementation detail outside the required scope of the RFC, so not worth holding anything up over.

Author: OK, I'll take the liberty to streamline that code if necessary.

invoked in order to start the chain runtime.

> When a channel is first joined with a config block, the channel's ledger folder is created, and inside it the config 

Reviewer 4: This seems like it might be a bit racy -- we could crash after creating the ledger folder but before writing the block.  Perhaps it would be better to reverse the order and locate the block outside the ledger dir so that we could resume creating the ledger dir if interrupted?

Author: You are correct, this needs to be done atomically. One way to do this is to create a temporary folder with all the resources and then rename it into place. Renaming is atomic so if we fail before the rename nothing happens.

it will delete the `.join-config-block` file.

The process of a join includes fetching blocks from other OSNs. For that purpose we will reuse code from 
`orderer/common/cluster`, in particular the method that pulls a single named channel 
`Replicator.PullChannel(channel string)`.

Deleting a channel is achieved by first adding to the ledger folder a file called `.remove`, again, to mark it for 
removal across restarts. Then, the runtime components of the channel are terminated (using `.../multichannel`),
and finally, if respective flag is set, the folder is deleted.

Listing the channel or channel is achieved by querying the channel map in the `.../channelparticipation` package 
implementation.

The interface of executing channel participation commands is expected to stay stable even if the  authentication and 
authorization mechanisms change, or even if this API is separated on to a different service endpoint.

## Boot sequence of Orderer
In the boot sequence of the orderer (`orderer/server/main.go`) we will have to check whether the orderer is working
with or without the system channel, and then start either the current on-boarding code (eventually calling 

> `orderer/common/cluster/Replicator`) or the new channel participation component.

Reviewer 4: I was initially thrown for a loop here, but I think you mean the component which implements the following/joining go routines, not the the participation service which still allows for the deletion of the system channel and listing of channels.  Could be clarified.

Author: I clarified the text.
